Docs: refactor and integrate into ROCm docs portal

peterjunpark opened the pull request:
This PR aligns Omniperf's documentation with other ROCm components for publication on rocm.docs.amd.com.
- Move docs/ to root.
- Update Sphinx build config.
- Structure docs for Diataxis.
- Convert to rST.
- Improve wording in some areas.
- Change client-side/server-side to core install/grafana server setup
- Add linting for docs

dgaliffiAMD reviewed (changes requested):
Please include a "Signed-off-by:" message in your commits. 
You can read more about how to resolve this in the DCO Check message: https://github.com/ROCm/omniperf/pull/362/checks

peterjunpark commented on docs/concept/command-processor.rst:
will move to [rocm-docs-core](https://github.com/ROCm/rocm-docs-core) instead

peterjunpark commented:
will add .wordlist.txt in https://github.com/ROCm/rocm-docs-core instead of here

peterjunpark commented:
still some formatting issues to fix. Mostly in `Profiling by example` and `Performance model`

peterjunpark commented:
demo: https://advanced-micro-devices-demo--362.com.readthedocs.build/projects/omniperf/en/362/

coleramos425 reviewed (approved):
This looks great @peterjunpark, nice job! I've read through the new docs using the demo link provided. Everything looks good to me.

coleramos425 commented:
Could we land this on `dev` before upstreaming to `main`? This way we're being consistent with [Contributing.md](https://github.com/ROCm/omniperf/blob/main/CONTRIBUTING.md) and we can include these with other bug fixes for the next release.

skyreflectedinmirrors commented:
I don't see the performance model in the demo link?

peterjunpark commented:
> I don't see the performance model in the demo link?

My bad -- fixed.

skyreflectedinmirrors commented:
No worries, I'm about halfway through -- have a few comments but it's mostly looking very good

skyreflectedinmirrors commented on docs/conceptual/l2-cache.rst:
```diff
@@ -0,0 +1,763 @@
+
+     - The ratio of the total number of cycles the L2-Fabric interface was
+       stalled on a read request to any destination (local HBM, remote PCIe
+       connected accelerator or CPU, or remote Infinity Fabric connected
```
Likely should have the (r) on PCIe, as it's the first use.  Again see note about (tm) for instinct

peterjunpark replied:
done

skyreflectedinmirrors commented on docs/conceptual/l2-cache.rst:
```diff
@@ -0,0 +1,763 @@
+   technology can be used to connect multiple accelerators to achieve advanced
+   peer-to-peer connectivity and enhanced bandwidths over traditional PCIe
+   connections. Some AMD Instinct MI accelerators like the MI250X,
+   `feature coherent CPU↔accelerator connections built using AMD Infinity Fabric <https://www.amd.com/system/files/documents/amd-cdna2-white-paper.pdf>`_
```
Missing a period to end the sentence.

peterjunpark replied:
fixed

skyreflectedinmirrors reviewed (changes requested):
Through the L2 cache, still need to cover SE-on

peterjunpark commented on docs/conceptual/performance-model.rst:
```diff
@@ -0,0 +1,76 @@
+
+   In this chapter, **MI2XX** refers to any of the CDNA2 architecture-based AMD
+   Instinct MI250X, MI250, and MI210 accelerators interchangeably in cases
+   where the exact product at hand is not vital.
```
Yeah, I agree with you actually.

peterjunpark replied:
fixed

lpaoletti commented on docs/index.rst:
```diff
@@ -0,0 +1,87 @@
+   :keywords: Omniperf, ROCm, profiler, tool, Instinct, accelerator, AMD
+
+********
+Omniperf
```
```suggestion
Omniperf documentation
```

peterjunpark replied:
updated

lpaoletti commented on docs/index.rst:
```diff
@@ -0,0 +1,87 @@
+documentation also explains the ideas motivating the design behind the tool and
+its components.
+
+Omniperf is open source and hosted at `<https://github.com/ROCm/omniperf>`__.
```
This paragraph should be below the next paragraph

peterjunpark replied:
updated

dgaliffiAMD commented on docs/install/core-install.rst:
```diff
@@ -0,0 +1,201 @@
+
+Installation steps
+------------------
+
```
Hi @peterjunpark , 

What are your thoughts about adding a subsection for "installing via package", which should be possible after rocm-6.2. 
Assuming rocm-6.2 is already installed, it should be something like this: 
```
sudo apt install omniperf
pip install -r /opt/rocm/libexec/omniperf/requirements.txt
```
Replace `apt` with the appropriate package installer for your OS.

peterjunpark replied:
Thanks @dgaliffiAMD. Added a thing.

Could you review this? https://advanced-micro-devices-demo--362.com.readthedocs.build/projects/omniperf/en/362/install/core-install.html#install-via-package-manager

Let me know if there are any mistakes or if you have any other suggestions. Thanks!

dgaliffiAMD commented on docs/install/core-install.rst:
```diff
@@ -0,0 +1,201 @@
+Omniperf relies on :doc:`ROCProfiler <rocprofiler:index>`'s ``rocprof`` binary
+during the profiling process. Normally, the path to this binary is detected
+automatically, but you can override the path by the setting the optional
+``ROCPROF`` environment variable.
```
I think it would be important to note that setting ROCPROF may be essential to running omniperf successfully. 
For example, MI300 will require `export ROCPROC=rocprofv2` otherwise you'll see the following error:   `ERROR gfx942 is not enabled in rocprofv1. Available profilers include: ['rocprofv2']`

skyreflectedinmirrors replied:
We should probably just relax this requirement now.  My understanding from the profiler team is gfx942 is fully supported in v1 at this point.

coleramos425 replied:
Agreed. I've filed a ticket to track progress on getting this change made:
https://github.com/ROCm/omniperf/issues/388

@nartmada could you assign this to a developer as you see fit? Thank you.

peterjunpark commented on docs/conceptual/performance-model.rst:
```diff
@@ -0,0 +1,76 @@
+hardware blocks of AMD Instinct accelerators. This section describes each
+hardware block on the accelerator as interacted with by a software developer to
+give a deeper understanding of the metrics reported by profiling data. Refer to
+:doc:`../how-to/profile/mode` for more practical examples and details on how
```
fixed

peterjunpark commented on docs/conceptual/performance-model.rst:
```diff
@@ -0,0 +1,76 @@
+   :prod-page:`MI250 <mi200/mi250>`, and :prod-page:`MI210 <mi200/mi210>`
+   product pages.
+
+In this chapter, the way Omniperf models performance is explained in detail
```
fixed

peterjunpark commented on docs/conceptual/compute-unit.rst:
```diff
@@ -0,0 +1,53 @@
+execution pipelines -- VALU, SALU, LDS, scheduler, and so forth. The metrics
+presented by Omniperf for these pipelines are described in
+:doc:`pipeline-metrics`. The :doc:`vL1D <vector-l1-cache>` cache and
+:doc:`LDS <local-data-share>` are described their own chapters.
```
fixed

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+On CDNA accelerators, the LDS contains 32 banks and each bank is 4B wide.
+The LDS is designed such that each bank can be read from, written to, or
+atomically updated every cycle, for a total throughput of 128B/clock
+:gcn-crash-course:`40`.
```
fixed

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+no conflicts in any single cycle.
+
+When multiple work-items want to read from the same address within a bank, the
+result can be efficiently broadcasted :gcn-crash-course:`41`. Multiple
```
fixed

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+When multiple work-items want to read from the same address within a bank, the
+result can be efficiently broadcasted :gcn-crash-course:`41`. Multiple
+work-items writing to the same address within a bank typically results undefined
+behavior in HIP and other languages, as the LDS will write the value from the
```
fixed

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+result can be efficiently broadcasted :gcn-crash-course:`41`. Multiple
+work-items writing to the same address within a bank typically results undefined
+behavior in HIP and other languages, as the LDS will write the value from the
+last work-item as determined by the hardware scheduler :gcn-crash-course:`41`.
```
fixed

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+
+* Issues up to one instruction per each of the instruction categories among the waves on the selected SIMD:
+
+  * :ref:`VALU <desc-valu>` / :ref:`VMEM <desc-valu>` operations
```
fixed. thanks

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+            GPRs required for D: 4
+            GPR alignment requirement: 8 bytes
+
+For the purposes of Omniperf, the MFMA unit is typically treated as a separate
```
added *typically*

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+   The exact details of VALU and MFMA operation co-execution vary by
+   instruction, and can be explored in more detail via the following fields in
+   the
+   `AMD Matrix Instruction Calculator's detailed instruction information <https://github.com/ROCm/amd_matrix_instruction_calculator#example-of-querying-instruction-information>`_.
```
fixed

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+^^^^^^^
+
+Barriers are resources on the compute-unit of a CDNA accelerator that
+are used to implement synchronization primitives; for example, HIP’s
```
fixed

peterjunpark commented on docs/conceptual/pipeline-metrics.rst:
```diff
@@ -0,0 +1,903 @@
+
+     - The total number of vector arithmetic logic unit (VALU) operations
+       issued. These are the workhorses of the
+       :doc:`compute unit <compute-unit>`, and are used to execute wide range of
```
fixed

peterjunpark commented on docs/conceptual/performance-model.rst:
```diff
@@ -0,0 +1,76 @@
+
+.. _perf-model-ext-refs:
+
+References
```
updated

peterjunpark commented on docs/conceptual/pipeline-metrics.rst:
```diff
@@ -0,0 +1,903 @@
+
+This section breaks down the types of vector memory (VMEM) instructions
+that were issued. Refer to the
+`Instruction Counts metrics section <desc-ta>` of address-processor front end of
```
fixed

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+
+   * - Workgroup manager → Data-return Stall
+
+     - Percent of the :ref:`total CU cycles <total-cu-cycles>` the data-return unit was stalled by the [workgroup manager](SPI) due to initialization of registers as a part of launching new workgroups.
```
fixed

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+The data-return path of the vL1D cache, also known as the Texture Data
+(TD) unit, is responsible for routing data returned from the
+:ref:`vL1D cache RAM <desc-tc>` back to a wavefront on a SIMD. As described in
+the `vL1D cache front-end <desc-ta>` section, the data-return path is passed
```
fixed

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+
+   * - L1-L2 Writes
+
+     - The number of post-coalescing write requests that are sent through the
```
fixed

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+
+     - Cache lines
+
+   * - Cache Hits
```
fixed

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+
+   * - Cache Hit Rate
+
+     - The ratio of the number of vL1D cache line requests that hit in vL1D
```
fixed

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+  Interface <TCP_TCC_Transactions_Detail>`__ for misses or the
+  :ref:`cache RAM <desc-tc>` for hits.
+
+* The cache RAM, also known as the `texture cache (TC) <desc-tc>`, stores
```
fixed

peterjunpark commented on docs/conceptual/pipeline-metrics.rst:
```diff
@@ -0,0 +1,903 @@
+:ref:`VALU <desc-valu>` and :ref:`MFMA <desc-mfma>` operations of the same precision
+(e.g., F32) in the same metric. Additionally, this panel lets the user
+control how the data is normalized (i.e., control the
+`normalization unit <normalization-units>`), while the speed-of-light panel does
```
fixed

peterjunpark commented on docs/conceptual/pipeline-metrics.rst:
```diff
@@ -0,0 +1,903 @@
+   other instruction types (namely: ``s_endpgm``) that are *typically* a very
+   small percentage of the overall kernel execution. This complication is
+   omitted for simplicity, but may result in small amounts of branch utilization
+   (less than 1%) for otherwise branch-less kernels.
```
fixed

peterjunpark commented on docs/conceptual/pipeline-metrics.rst:
```diff
@@ -0,0 +1,903 @@
+       kernel in cycles. Computed as the ratio of the total number of cycles the
+       MFMA unit was busy over the total number of MFMA instructions. Compare
+       to, for example, the
+       `AMD Matrix Instruction Calculator <https://github.com/RadeonOpenCompute/amd_matrix_instruction_calculator>`.
```
fixed

peterjunpark commented on docs/conceptual/pipeline-metrics.rst:
```diff
@@ -0,0 +1,903 @@
+
+.. _mfma-instruction-mix:
+
+MFMA instruction mix
```
fixed

peterjunpark commented on docs/conceptual/pipeline-metrics.rst:
```diff
@@ -0,0 +1,903 @@
+This section reports a number of key performance characteristics of
+various execution units on the :doc:`CU <compute-unit>`. Refer to
+:ref:`ipc-example` for a detailed dive into these metrics, and
+:ref:`scheduler <desc-scheduler>` for a high-level overview of execution units
```
fixed

peterjunpark commented on docs/conceptual/performance-model.rst:
```diff
@@ -0,0 +1,76 @@
+
+* :llvm-docs:`LLVM User Guide for AMDGPU Backend <>`
+
+.. rubric:: Disclaimer
```
moved to relevant sections

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+Together, this complex is known as the vL1D, or Texture Cache per Pipe
+(TCP). A simplified diagram of the vL1D is presented below:
+
+.. figure:: ../data/performance-model/l1perf_model.*
```
fixed

peterjunpark commented on docs/conceptual/pipeline-descriptions.rst:
```diff
@@ -0,0 +1,292 @@
+(CDNA2) has a 512 KiB VGPR file per CU, where each wave can dynamically request
+up to 256 KiB of VGPRs and an additional 256 KiB of AGPRs. For more information,
+refer to 
+`<https://github.com/RadeonOpenCompute/ROCm/issues/1689#issuecomment-1553751913>`__.
```
updated

peterjunpark commented on docs/conceptual/pipeline-metrics.rst:
```diff
@@ -0,0 +1,903 @@
+       :ref:`normalization unit <normalization-units>`. Note: on current CDNA
+       accelerators, the VALU has no native BF16 instructions.
+
+     - FLOP per normalization unit
```
done everywhere

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+
+   * - Hit Rate
+
+     - The ratio of the number of vL1D cache line requests that hit [#vl1d-hit]_
```
fixed footnotes

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+
+     - Private memory, or "scratch" memory, is only visible to a particular
+       :ref:`work-item <desc-work-item>` in a particular
+       :ref:`workgroup <desc-workgroup>`. On AMD Instinct MI-series
```
Yeah, my understanding is to mark the first instance on a page.

peterjunpark commented on docs/conceptual/vector-l1-cache.rst:
```diff
@@ -0,0 +1,761 @@
+
+     - Indicates how well memory instructions were coalesced by the
+       :ref:`address processing unit <desc-ta>`, ranging from uncoalesced (25%)
+       to fully coalesced (100%). The average number of
```
updated

skyreflectedinmirrors commented on docs/conceptual/shader-engine.rst:
```diff
@@ -0,0 +1,704 @@
+-------------------
+
+This panel gives more detail on the data requested across the
+sL1D↔:doc:`L2 <l2-cache>` interface.
```
Link is broken

peterjunpark replied:
fixed

skyreflectedinmirrors commented on docs/conceptual/shader-engine.rst:
```diff
@@ -0,0 +1,704 @@
+   * - sL1D-L2 BW
+
+     - The total number of bytes read from, written to, or atomically updated
+       across the sL1D↔:doc:`L2 <l2-cache>` interface, per
```
Link is broken

peterjunpark replied:
fixed

skyreflectedinmirrors commented on docs/conceptual/shader-engine.rst:
```diff
@@ -0,0 +1,704 @@
+
+   * - Stall Cycles
+
+     - The total number of cycles the sL1D↔:doc:`L2 <l2-cache>` interface was
```
Link is broken

peterjunpark replied:
fixed

skyreflectedinmirrors commented on docs/conceptual/system-speed-of-light.rst:
```diff
@@ -0,0 +1,318 @@
+
+   The theoretical maximum throughput for some metrics in this section are
+   currently computed with the maximum achievable clock frequency, as reported
+   by :doc:`rocminfo:index`, for an accelerator. This may not be realistic for
```
it isn't reported by the rocminfo documentation, but by the rocminfo _tool_.  I would phrase this as "as reported by ...`rocminfo`, the documentation for which can be found ..."

peterjunpark replied:
fixed

skyreflectedinmirrors commented on docs/conceptual/includes/normalization-units.rst:
```diff
@@ -0,0 +1,45 @@
+
+By default, Omniperf uses the ``per_wave`` normalization.
+
+The ideal normalization varies depending on your use case. For instance, a
```
'best' normalization may vary depending...

peterjunpark replied:
fixed

peterjunpark commented on docs/conceptual/definitions.rst:
```diff
@@ -0,0 +1,109 @@
+
+.. _memory-type:
+
+Memory type
```
fixed

peterjunpark commented:
@skyreflectedinmirrors Thanks for the feedback ❤️. Everything you pointed out has been addressed. Pls let me know if there are any other glaring issues.

Also, what's the plan for the old version of the docs? Should we archive them or straight up remove them?

We need to get this into the release branch sooner rather than later.

skyreflectedinmirrors commented on docs/tutorial/includes/instructions-per-cycle-and-utilizations.rst:
```diff
@@ -0,0 +1,485 @@
+
+.. _issued-ipc:
+
+Exploring “issued” IPC via MFMA operations
```
This should be a header

peterjunpark replied:
7912d52

skyreflectedinmirrors reviewed (commented):
Found one more @peterjunpark

coleramos425 commented:
> Thanks for all your quick edits @peterjunpark. My 2 cents is to archive the old docs for a release or two, and then add a redirect from the old site to the new one (if it's going to be a different URL, anyways). But I'd be interested in what @coleramos425 has to say

Agreed. In the current `dev` branch we have a subdirectory (omniperf/src/archive/docs-x.x) such that the [docs workflow](https://github.com/peterjunpark/omniperf/blob/7912d5299c629b34a23866772646b2116821fe43/.github/workflows/docs.yml#L31-L52) will build archived versions of our docs (e.g. accommodating customers with older hyperlinks).

Feel free to move the directory around as you see fit @peterjunpark, but we should probably move the old docs in here and build this version of the docs from the archive subdirectory in a similar fashion.

peterjunpark commented:
> > Thanks for all your quick edits @peterjunpark. My 2 cents is to archive the old docs for a release or two, and then add a redirect from the old site to the new one (if it's going to be a different URL, anyways). But I'd be interested in what @coleramos425 has to say
> 
> Agreed. In the current `dev` branch we have a subdirectory (omniperf/src/archive/docs-x.x) such that the [docs workflow](https://github.com/peterjunpark/omniperf/blob/7912d5299c629b34a23866772646b2116821fe43/.github/workflows/docs.yml#L31-L52) will build archived versions of our docs (e.g. accommodating customers with older hyperlinks).
> 
> Feel free to move the directory around as you see fit @peterjunpark, but we should probably move the old docs in here and build this version of the docs from the archive subdirectory in a similar fashion.

@coleramos425 @skyreflectedinmirrors Done in [8480640](https://github.com/ROCm/omniperf/pull/362/commits/8480640709b213725f222b038417e83b49cda992), [85c27a8](https://github.com/ROCm/omniperf/pull/362/commits/85c27a849d5b85da395a53300a4bfa1deea6e978), and [13c64b2](https://github.com/ROCm/omniperf/pull/362/commits/13c64b24f85efff2d6c5d7c05baaf921d2afff71).

Moved it to `docs/archive/docs-2.x` alongside `docs/archive/docs-1.x`. Updated workflows and Sphinx confs accordingly.

LMK if there are any concerns. Thanks.
